From 3e426f6d4e9c3701fa5197fe0f12fcfe2389e362 Mon Sep 17 00:00:00 2001
From: James Le Cuirot <jlecuirot@microsoft.com>
Date: Fri, 4 Sep 2026 18:46:20 +0100
Subject: [PATCH 14/14] internal/resource/url: Add an optional oem:// schema to
 read from /oem

Ignition previously supported this, but it now requires less additional
code thanks to the existing file:// support.

This would have been added to Butane's Flatcar variant, but that only
covers use of oem:// as a resource, not as a ignition.config.url
location.

Signed-off-by: James Le Cuirot <jlecuirot@microsoft.com>

diff --git a/config/v3_7_experimental/types/url.go b/config/v3_7_experimental/types/url.go
index 97120e89..feed44d3 100644
--- a/config/v3_7_experimental/types/url.go
+++ b/config/v3_7_experimental/types/url.go
@@ -70,7 +70,7 @@ func validateURL(s string) error {
 			return err
 		}
 		return nil
-	case "file":
+	case "file", "oem":
 		if u.Host != "" || u.Path == "" {
 			return errors.ErrPathNotAbsolute
 		}
diff --git a/config/v3_7_experimental/types/url_test.go b/config/v3_7_experimental/types/url_test.go
index 1901b7af..7b014e15 100644
--- a/config/v3_7_experimental/types/url_test.go
+++ b/config/v3_7_experimental/types/url_test.go
@@ -126,6 +126,14 @@ func TestURLValidate(t *testing.T) {
 			util.StrToPtr("gs://bucket/object"),
 			nil,
 		},
+		{
+			util.StrToPtr("oem:///config.ign"),
+			nil,
+		},
+		{
+			util.StrToPtr("oem://config.ign"),
+			errors.ErrPathNotAbsolute,
+		},
 	}
 
 	for i, test := range tests {
diff --git a/internal/distro/distro.go b/internal/distro/distro.go
index 5e03425b..e375ed36 100644
--- a/internal/distro/distro.go
+++ b/internal/distro/distro.go
@@ -81,6 +81,10 @@ var (
 	// ".ssh/authorized_keys.d/ignition" ("true"), or to
 	// ".ssh/authorized_keys" ("false").
 	writeAuthorizedKeysFragment = "true"
+	// oemFetch indicates whether the "oem" URL scheme is supported,
+	// reading files from the local "/oem" directory. It is only
+	// meaningful on distros that provide this directory (e.g. Flatcar).
+	oemFetch = "false"
 
 	// Special file paths in the real root
 	luksRealRootKeyFilePath = "/etc/luks/"
@@ -150,6 +154,7 @@ func BlackboxTesting() bool { return bakedStringToBool(blackboxTesting) }
 func WriteAuthorizedKeysFragment() bool {
 	return bakedStringToBool(fromEnv("WRITE_AUTHORIZED_KEYS_FRAGMENT", writeAuthorizedKeysFragment))
 }
+func OEMFetch() bool { return bakedStringToBool(oemFetch) }
 
 func fromEnv(nameSuffix, defaultValue string) string {
 	value := os.Getenv("IGNITION_" + nameSuffix)
diff --git a/internal/resource/url.go b/internal/resource/url.go
index 63e9b7d1..41cf5cd8 100644
--- a/internal/resource/url.go
+++ b/internal/resource/url.go
@@ -35,6 +35,7 @@ import (
 
 	"cloud.google.com/go/compute/metadata"
 	configErrors "github.com/coreos/ignition/v2/config/shared/errors"
+	"github.com/coreos/ignition/v2/internal/distro"
 	"github.com/coreos/ignition/v2/internal/log"
 	"github.com/coreos/ignition/v2/internal/util"
 	"github.com/coreos/vcontext/report"
@@ -185,6 +186,11 @@ func (f *Fetcher) FetchToBuffer(u url.URL, opts FetchOptions) ([]byte, error) {
 		err = f.fetchFromGCS(u, dest, opts)
 	case "file":
 		err = f.fetchFromFile(u.Path, dest, opts)
+	case "oem":
+		if !distro.OEMFetch() {
+			return nil, ErrSchemeUnsupported
+		}
+		err = f.fetchFromFile(filepath.Join("/oem", u.Path), dest, opts)
 	case "":
 		return nil, nil
 	default:
@@ -255,6 +261,11 @@ func (f *Fetcher) Fetch(u url.URL, dest *os.File, opts FetchOptions) error {
 		return f.fetchFromGCS(u, dest, opts)
 	case "file":
 		return f.fetchFromFile(u.Path, dest, opts)
+	case "oem":
+		if !distro.OEMFetch() {
+			return ErrSchemeUnsupported
+		}
+		return f.fetchFromFile(filepath.Join("/oem", u.Path), dest, opts)
 	case "":
 		return nil
 	default:
diff --git a/internal/util/url.go b/internal/util/url.go
index de465ba4..63828c95 100644
--- a/internal/util/url.go
+++ b/internal/util/url.go
@@ -19,5 +19,5 @@ import (
 )
 
 func UrlNeedsNet(u url.URL) bool {
-	return u.Scheme != "data" && u.Scheme != "file" && u.Scheme != ""
+	return u.Scheme != "data" && u.Scheme != "file" && u.Scheme != "oem" && u.Scheme != ""
 }
-- 
2.55.0

